d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,709 @@
+      if (subscriber == null)
+      {
+        _log.info("Subscribe RDS resource {}", resourceName);
+        subscriber = new ResourceSubscriber(ResourceType.RDS, resourceName);
+        _rdsResourceSubscribers.put(resourceName, subscriber);
+
+        if (_adsStream == null)
+        {
+          startRpcStream();
+        }
+        _adsStream.sendDiscoveryRequest(ResourceType.RDS, Collections.singletonList(resourceName));
+      }
+      subscriber.addWatcher(watcher);
+    });
+  }
+
+  @Override
+  void watchCdsResource(final String resourceName, final CdsResourceWatcher watcher)
+  {
+    _executorService.execute(() ->
+    {
+      ResourceSubscriber subscriber = _cdsResourceSubscribers.get(resourceName);
+      if (subscriber == null)
+      {
+        _log.info("Subscribe CDS resource {}", resourceName);
+        subscriber = new ResourceSubscriber(ResourceType.CDS, resourceName);
+        _cdsResourceSubscribers.put(resourceName, subscriber);
+
+        if (_adsStream == null)
+        {
+          startRpcStream();
+        }
+        _adsStream.sendDiscoveryRequest(ResourceType.CDS, Collections.singletonList(resourceName));
+      }
+      subscriber.addWatcher(watcher);
+    });
+  }
+
+  @Override
+  void watchEdsResource(final String resourceName, final EdsResourceWatcher watcher)
+  {
+    _executorService.execute(() ->
+    {
+      ResourceSubscriber subscriber = _edsResourceSubscribers.get(resourceName);
+      if (subscriber == null)
+      {
+        _log.info("Subscribe EDS resource {}", resourceName);
+        subscriber = new ResourceSubscriber(ResourceType.EDS, resourceName);
+        _edsResourceSubscribers.put(resourceName, subscriber);
+
+        if (_adsStream == null)
+        {
+          startRpcStream();
+        }
+        _adsStream.sendDiscoveryRequest(ResourceType.EDS, Collections.singletonList(resourceName));
+      }
+      subscriber.addWatcher(watcher);
+    });
+  }
+
+  private void handleLdsResponse(DiscoveryResponseData responseData)
+  {
+    Map<String, LdsUpdate> ldsUpdates = new HashMap<>();
+    List<String> errors = new ArrayList<>();
+
+    for (Resource resource : responseData.getResourcesList())
+    {
+      try
+      {
+        Listener listener = resource.getResource().unpack(Listener.class);
+
+        for (FilterChain filterChain : listener.getFilterChainsList())
+        {
+          for (Filter filter : filterChain.getFiltersList())
+          {
+            if (filter.hasTypedConfig() && filter.getTypedConfig()
+                .getTypeUrl()
+                .equals(TYPE_URL_HTTP_CONNECTION_MANAGER))
+            {
+              HttpConnectionManager httpConnectionManager = filter.getTypedConfig().unpack(HttpConnectionManager.class);
+              ldsUpdates.put(listener.getName(), new LdsUpdate(httpConnectionManager));
+            }
+          }
+        }
+      } catch (InvalidProtocolBufferException e)
+      {
+        _log.warn("Failed to unpack LDS response", e);
```

Comment:
Helpful to print the listener proto

Reply:
The problem is that this exception is thrown by Resource.unpack(). If the resource is malformed, it cannot be translated into a listener proto. Hopefully, this error message e contains all information needed.